projects/vdk-plugins/vdk-data-sources/src/vdk/plugin/data_sources/ingester.py

```diff
@@ -0,0 +1,247 @@
+
+    @staticmethod
+    def _is_termination_signal(queue_item):
+        return queue_item is None
+
+    def _ingest_stream(self, ingest_entry: IngestQueueEntry):
+        for payload in ingest_entry.stream.read():
+            log.debug(f"Ingest payload {payload}")
+            if not payload.data and not payload.state:
+                log.warning(
+                    f"{ingest_entry.stream.name()} returned payload without any data or state. "
+                    f"This is pretty meaningless so we will skip it. But it is suspicious."
+                )
+                continue
+
+            if payload.data:
+                for destination in ingest_entry.destinations:
+                    destination_table = self._infer_destination_table(
+                        ingest_entry, destination, payload
+                    )
+
+                    payload_to_sent = DataSourcePayload(
+                        payload.data, payload.metadata, payload.state, destination_table
+                    )
+                    if destination.map_function:
+                        payload_to_sent = destination.map_function(payload_to_sent)
+                        if not payload_to_sent:
+                            continue
+
+                    self.__actual_ingester.send_object_for_ingestion(
+                        payload=payload_to_sent.data,
+                        destination_table=payload_to_sent.destination_table,
+                        method=destination.method,
+                        target=destination.target,
+                        collection_id=destination.collection_id,
+                    )
+
+            if payload.state:
+                # TODO: it would be better to update the state after the actual ingester
+                # is finished. So it might be good idea to add a callback so that when send object for ingestion
+                # finishes successfully then state is updated.
+                data_source_state = self.__state_factory.get_data_source_state(
+                    ingest_entry.data_source_id
+                )
+                data_source_state.update_stream(
+                    ingest_entry.stream.name(), payload.state
+                )
+
+    @staticmethod
+    def _infer_destination_table(
+        ingest_entry: IngestQueueEntry, destination: IngestDestination, payload
+    ):
+        if destination.destination_table:
+            destination_table = destination.destination_table
+        elif payload.destination_table:
+            destination_table = payload.destination_table
+        else:
+            destination_table = ingest_entry.stream.name()
+        return destination_table
+
+    def start_ingestion(
+        self,
+        data_source_id: str,
+        data_source: IDataSource,
+        destinations: List[IngestDestination] = None,
+        error_callback: Optional[IDataSourceErrorCallback] = None,
+    ):
+        if data_source_id not in self.__ingested_streams_set:
```

Comment:
"ingested" streams? What if the ingestion fails? I would rename it to something something like "being_ingested" or "ingesting" streams?

Reply:
Sounds good